src/pykeen/utils.py

```diff
@@ -1095,3 +1100,144 @@ def load_configuration(path: Union[str, pathlib.Path, os.PathLike]) -> Mapping[s
     import doctest
 
     doctest.testmod()
+
+
+def product_normalize(x: torch.FloatTensor, dim: int = -1) -> torch.FloatTensor:
+    r"""Normalize a tensor along a given dimension so that the geometric mean is 1.0.
+
+    :param x: shape: s
+        An input tensor
+    :param dim:
+        the dimension along which to normalize the tensor
+
+    :return: shape: s
+        An output tensor where the given dimension is normalized to have a geometric mean of 1.0.
+    """
+    return x / (x.abs().clamp(min=SANITY_EPSILON).log().mean(dim=dim, keepdim=True).exp().clamp(min=SANITY_EPSILON))
+
+
+def compute_box(
+    base: torch.FloatTensor,
+    delta: torch.FloatTensor,
+    size: torch.FloatTensor,
+) -> Tuple[torch.FloatTensor, torch.FloatTensor]:
+    r"""Compute the lower and upper corners of a resulting box.
+
+    :param base: shape: (*, d)
+        the base position (box center) of the input relation embeddings
+    :param delta:  shape: (*, d)
+        the base shape of the input relation embeddings
+    :param size: shape: (*, d)
+        the size scalar vectors of the input relation embeddings
+
+    :return: shape: (*, d) each
+        lower and upper bounds of the box whose embeddings are provided as input.
+    """
+    # Enforce that sizes are strictly positive by passing through ELU
+    size_pos = torch.nn.functional.elu(size) + 1
+
+    # Shape vector is normalized using the above helper function
+    delta_norm = product_normalize(delta)
+
+    # Size is learned separately and applied to normalized shape
+    delta_final = size_pos * delta_norm
+
+    # Compute potential boundaries by applying the shape in substraction
+    first_bound = base - 0.5 * delta_final
+
+    # and in addition
+    second_bound = base + 0.5 * delta_final
+
+    # Compute box upper bounds using min and max respectively
+    box_low = torch.minimum(first_bound, second_bound)
```

Comment:
would it be more simple just to check if base < 0 and assign these in the conditional?

Reply:
No, since this operation happens in parallel / vectorized, whereas a "normal" conditional would either require conversion from torch to Python (bad, in particular since the needs to be transferred to cpu!), or usage of torch.where, but this should be roughly the same speed.